### PR TITLE
TINY-7619: Ensure all dialogs have the ability to be opened by a command

### PR DESCRIPTION
**Pull request by diogo405**
Related Ticket:

Description of Changes:
* Changed emoticons, wordcount, code, and codesample plugins to open dialogs using commands

Pre-checks:
* [x] Changelog entry added
* [x] ~Tests have been added (if applicable)~
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] ~Review comments resolved~

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/CHANGELOG.md [hunk 1]
```diff
@@ -29,7 +29,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - The context toolbar will now move out of the way when it overlaps with the selection, such as in table cells #TINY-7192
 
 ### Changed
-- Changed the load order so that the content css gets loaded before the editor gets populated with contents. #TINY-7249
+- Changed the load order so that the content css gets loaded before the editor gets populated with contents #TINY-7249
+- Changed emoticons, wordcount, code, and codesample plugins to open dialogs using commands #TINY-7619
```
Comment (lnewson) [inline modules/tinymce/CHANGELOG.md:33]: This also needs to mention that we have the 2 new commands: `mceEmoticons` and `mceWordCount`. However, that should be under the `Added` section instead, so we'll need 2 changelog entries here.

- [ ] Add a changelog entry for the new commands

File: modules/tinymce/src/plugins/codesample/main/ts/ui/Buttons.ts [hunk 2]
```diff
@@ -18,7 +16,7 @@ const register = (editor: Editor) => {
   editor.ui.registry.addToggleButton('codesample', {
     icon: 'code-sample',
     tooltip: 'Insert/edit code sample',
-    onAction: () => Dialog.open(editor),
+    onAction: () => editor.execCommand('codesample'),
```
Comment (lnewson) [inline modules/tinymce/src/plugins/codesample/main/ts/ui/Buttons.ts:19]: Don't forgot to also clean up the duplication with `onAction` here as well.

File: modules/tinymce/src/plugins/wordcount/main/ts/ui/Buttons.ts [hunk 3]
```diff
@@ -7,20 +7,17 @@
...
-import { WordCountApi } from '../api/Api';
-import * as Dialog from './Dialog';
-
-const register = (editor: Editor, api: WordCountApi) => {
+const register = (editor: Editor) => {
   editor.ui.registry.addButton('wordcount', {
     tooltip: 'Word count',
     icon: 'character-count',
-    onAction: () => Dialog.open(editor, api)
+    onAction: () => editor.execCommand('mceWordCount')
```
Comment (lnewson) [inline modules/tinymce/src/plugins/wordcount/main/ts/ui/Buttons.ts:14]: Don't forgot to also clean up the duplication with `onAction` here as well.

File: modules/tinymce/src/plugins/code/main/ts/ui/Buttons.ts [hunk 4]
```diff
@@ -7,19 +7,17 @@
...
 import Editor from 'tinymce/core/api/Editor';
 
-import * as Dialog from './Dialog';
-
 const register = (editor: Editor) => {
   editor.ui.registry.addButton('code', {
     icon: 'sourcecode',
     tooltip: 'Source code',
-    onAction: () => Dialog.open(editor)
+    onAction: () => editor.execCommand('mceCodeEditor')
```
Comment (lnewson) [inline modules/tinymce/src/plugins/code/main/ts/ui/Buttons.ts:14]: I know this was like this before, but we have some duplication here. Can you please abstract this out to an `onAction` function similar to what was done in emoticons. This also ensures the actions are the same between toolbar button and menu item.

- [ ] Remove onAction duplication

**Review by lnewson — APPROVED** [review]

**Comment by lnewson** [comment]
I was also investigating to make sure we'd caught all of them and I noticed you've missed the `template` plugin as it currently doesn't have a command to open the dialog. So please make sure to add that and update the buttons to use the comand.

- [ ] Ensure the template plugin has a command to open the dialog

File: modules/tinymce/src/plugins/template/main/ts/api/Commands.ts [hunk 5]
```diff
@@ -7,10 +7,20 @@
...
+
+const showDialog = (editor: Editor) => {
+  return (templates) => {
+    Dialog.open(editor, templates);
+  };
+};
 
 const register = (editor) => {
   editor.addCommand('mceInsertTemplate', Fun.curry(Templates.insertTemplate, editor));
+  editor.addCommand('mceCreateTemplateList', Templates.createTemplateList(editor, showDialog(editor)));
```
Comment (lnewson) [inline modules/tinymce/src/plugins/template/main/ts/api/Commands.ts:23]: This doesn't match naming conventions and is inconsistent with other plugin commands. It should be `mceTemplate` in this case, so can we please rename that?

- [ ] Rename command to keep it consistent

Comment (diogo405) [reply modules/tinymce/src/plugins/template/main/ts/api/Commands.ts]: Changed to `mceTemplate` but couldn't figure out the convention. Is it mce<PLUGIN_FOLDER> capitalized?

Comment (lnewson) [reply modules/tinymce/src/plugins/template/main/ts/api/Commands.ts]: Yeah, the convention is normally just `mce<PLUGIN NAME>` where the plugin name uses PascalCase. There are some older things that don't follow that, such as `mceCodeEditor`, but we can't change them unfortunately.

File: modules/tinymce/CHANGELOG.md [hunk 6]
```diff
@@ -21,6 +21,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added a new `initData` property to `fancymenuitem` to allow custom initialization data #TINY-7480
 - Added a new `language` menu item and toolbar button to add `lang` attributes to content, with an accompanying `content_langs` setting to specify the languages available #TINY-6149
 - A new `lang` format is now available that can be used with `editor.formatter`, or applied with the `Lang` editor command #TINY-6149
+- Added new commands: `mceEmoticons` (emoticons plugin), `mceWordCount` (wordcount), and `mceTemplate` (template) #TINY-7619
```
Comment (lnewson) [inline modules/tinymce/CHANGELOG.md:24]:
```suggestion
- Added new plugin commands: `mceEmoticons` (emoticons), `mceWordCount` (wordcount), and `mceTemplate` (template) #TINY-7619
```
For consistency and to clarify they all are plugins.

**Review by lnewson — APPROVED** [review]

**Review by ashfordneil — APPROVED** [review]

**Review by ltrouton — APPROVED** [review]

File: modules/tinymce/CHANGELOG.md [hunk 7]
```diff
@@ -29,7 +30,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - The context toolbar will now move out of the way when it overlaps with the selection, such as in table cells #TINY-7192
 
 ### Changed
-- Changed the load order so that the content css gets loaded before the editor gets populated with contents. #TINY-7249
+- Changed the load order so that the content css gets loaded before the editor gets populated with contents #TINY-7249
+- Changed emoticons, wordcount, code, codesample, and template plugins to open dialogs using commands #TINY-7619
```
[the remark this answers is not in the document]
Comment (JamesToohey) [reply modules/tinymce/CHANGELOG.md:34]: If they're in the lower case format we should use backticks

[the remark this answers is not in the document]
Comment (JamesToohey) [reply modules/tinymce/CHANGELOG.md:34]:
```suggestion
- Changed `emoticons`, `wordcount`, `code`, `codesample`, and `template` plugins to open dialogs using commands #TINY-7619
```

[the remark this answers is not in the document]
Comment (diogo405) [reply modules/tinymce/CHANGELOG.md]: Looking at the CHANGELOG.md I think they should have. Just added.

**Review by metricjs — APPROVED** [review]
This is much cleaner, thanks Diogo!

**Pull request merged**